File: algorithm/python/FlattenAMultilevelDoublyLinkedList.py
# Definition for a Node.

class Node:
    def __init__(self, val, prev, next, child):
          self.val = val
          self.prev = prev
          self.next = next
          self.child = child

class Solution:
    def flatten(self, head):
        if not head: return

        dummy = Node(0, None, head, None)
        self.dfs(dummy, head)

        dummy.next.prev = None
        return dummy.next

    def dfs(self, prev, curr):
        if not curr:
          return prev
        
        curr.prev = prev
        prev.next = curr

        tempNext = curr.next
        tail = self.dfs(curr, curr.child)
        curr.child = None
        return self.dfs(tail, tempNext)

File: algorithm/Java/FlattenAMultilevelDoublyLinkedList.java
// Definition for a Node.
class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
};

class Solution {
    public Node flatten(Node head) {
        if (head == null) {
            return head;
        }
        Node dummy = new Node(0);
        dummy.next = head;
        dfs(dummy, head);
        dummy.next.prev = null;
        return dummy.next;
    }
    
    public Node dfs(Node prev, Node curr) {
        if (curr == null) {
            return prev;
        } 

        curr.prev = prev;
        prev.next = curr;

        Node tempNext = curr.next;
        Node tail = dfs(curr, curr.child);
        curr.child = null;
        return dfs(tail, tempNext);
    }
}